src/http.cc

```diff
@@ -1933,6 +2071,49 @@ HttpStateData::httpBuildRequestHeader(HttpRequest * request,
     strConnection.clean();
 }
 
+void
+HttpStateData::makeUpgradeHeaders(HttpHeader &hdr_out)
+{
+    // We need virgin request received by squid. This is stored to
+    // AccessLogEntry::request member.
+    const HttpHeader &hdr_in = fwd->al->request ? fwd->al->request->header :request->header;
```

Comment:
If I am not wrong it is not possible to have FwdState object without AccessLogEntry (any more) in squid. But probably it is better to check that fwd->al is not null. In other places we are checking it before using it.
I add it in check, I am not using assertion.

> Also, should we always use the virgin client Upgrade header (i.e., fwd->al->request)?

What do you mean? This is what it is used if the info is available.
Also if the virgin request is not adapted (ICAP/ECAP allow 204, or no ICAP/ECAP) the HttpStateData::request is the virgin.

@rousskov 
The UPGRADE header is missing from adapted HttpRequests. This is because removed from headers before sent to ICAP/ECAP. Should we copy it, silently, to adapted HttpRequest after this is received from ICAP/ECAP service?

Reply:
(Partially) addressed at d55c31a9

Reply:
> This is because removed from headers before sent to ICAP/ECAP.

I do not think we should be removing Upgrade. Instead, we should let the adaptation service decide whether an Upgrade is allowed and then honor its decision.

Yes, I know that ICAP [specs](https://tools.ietf.org/html/rfc3507#section-4.4.2) prohibit sending hop-by-hop headers to adaptation services. I believe that requirement itself was a protocol design mistake because adaptation services are used not just for end-to-end content adaptation but for access control, logging, and other hop-specific needs. I believe that requirement should be violated in this specific case.

Once the above is fixed, the problem/ambiguity discussed in this change request will be gone, right? The will just use the adapted request for all Upgrade-related decisions.

Reply:
I fully agree that at least some of the hop by hop headers must be sent to ICAP. There are some of them which maybe should not sent for example "Proxy-Authenticate" header (which may include info which must not go out of squid on a remote server), but I am not seeing any reason why the other hopbyhop headers to not be sent to ICAP. (If I am not wrong ECAP does not have such restrictions)

> Once the above is fixed, the problem/ambiguity discussed in this change request will be gone, right? The will just use the adapted request for all Upgrade-related decisions.

Yes.

Should we fix squid to allow all hop-by-hop headers to be sent to ICAP into this PR?
Actually inside Adaptation::Icap::ModXact::encapsulateHead there is the call 'headClone->header.removeHopByHopEntries();'. We can just remove this call, or just exclude Upgrade header (add it again after removeHopByHopEntries()).

Reply:
> Should we fix squid to allow all hop-by-hop headers to be sent to ICAP into this PR?

Or you may consider adding a TODO about this change and (for now) unconditionally use virgin request here (fwd->al->request).

Reply:
FTR: Hop-by-hop headers were hidden from ICAP servers by us, apparently to comply with the ICAP protocol requirement: https://bugs.squid-cache.org/show_bug.cgi?id=1923

> Should we fix squid to allow all hop-by-hop headers to be sent to ICAP into this PR?

I would be worried about introducing such a big change for no compelling reason. It may break buggy ICAP servers that are not ready to see `Connection` or some other hop-by-hop header there. It might also break Squid when it suddenly starts seeing some special hop-by-hop headers in the post-adaptation response.

I recommend against this allow/preserve-all change. I would wait for folks to ask to preserve specific hop-by-hop headers they need (like it happened here with Upgrade).

> or just exclude Upgrade header (add it again after removeHopByHopEntries()

I think this removing and adding is acceptable, at least until we have a second header that needs this special treatment (see above). At that point, we will probably add a removeHopByHopEntries() parameter that will specify excluded headers.

I think Upgrade should be preserved in this PR because the post-adaptation code should use adapted request (and temporary using the virgin request creates problems as discussed in this change request).

Let's hope no ICAP server filters Upgrade out just because it can be considered a hop-by-hop header!

Reply:
resolved at commit a30030c9